src/test/kotlin/cacophony/controlflow/generation/StructTest.kt

```diff
@@ -145,32 +132,298 @@ class StructTest {
+    }
+
+    @Test
+    fun `assign switched simple`() {
+        // given
+
+        /*
+         * let f = [] -> Unit => (let x = {a = 7, b = true}; x = {b = false, a = 5};);
+         */
+        val fDef =
+            unitFunctionDefinition(
+                "f",
+                listOf(),
+                block(
+                    variableDeclaration("x", simpleStruct()),
+                    variableUse("x") assign struct("b" to lit(false), "a" to lit(5)),
+                    empty(),
+                ),
+            )
+
+        // when
+        val actualCFG = generateSimplifiedCFG(fDef)
+
+        // then
+        val virA = Register.VirtualRegister()
+        val virB = Register.VirtualRegister()
+
+        val expectedCFG =
+            singleWrappedFragmentCFG(fDef) {
+                "bodyEntry" does jump("assign b") { writeRegister(virA, integer(7)) }
+                "assign b" does jump("reassign a") { writeRegister(virB, integer(1)) }
+                "reassign a" does jump("reassign b") { writeRegister(virA, integer(5)) }
+                "reassign b" does jump("assign res") { writeRegister(virB, integer(0)) }
+                "assign res" does jump("bodyExit") { writeRegister(getResultRegister(), integer(42)) }
+            }
+
+        assertEquivalent(actualCFG, expectedCFG)
+    }
+
+    @Test
+    fun `reassign in nested`() {
+        // given
+
+        /*
+         * let f = [] -> Int => (let x = {a = 5, b = {a = 7, b = true} }; x.b = {a = 5, b = false}; x.b.a);
+         */
+        val fDef =
+            intFunctionDefinition(
+                "f",
+                listOf(),
+                block(
+                    variableDeclaration("x", nestedStruct()),
+                    lvalueFieldRef(variableUse("x"), "b") assign simpleStruct(5, false),
+                    rvalueFieldRef(rvalueFieldRef(variableUse("x"), "b"), "a"),
+                ),
+            )
+
+        // when
+        val actualCFG = generateSimplifiedCFG(fDef)
+
+        // then
+        val virA = Register.VirtualRegister()
+        val virB = Register.VirtualRegister()
+        val virC = Register.VirtualRegister()
+
+        val expectedCFG =
+            singleWrappedFragmentCFG(fDef) {
+                "bodyEntry" does jump("assign b.a") { writeRegister(virA, integer(5)) }
+                "assign b.a" does jump("assign b.b") { writeRegister(virB, integer(7)) }
+                "assign b.b" does jump("reassign b.a") { writeRegister(virC, integer(1)) }
+                "reassign b.a" does jump("reassign b.b") { writeRegister(virB, integer(5)) }
+                "reassign b.b" does jump("assign res") { writeRegister(virC, integer(0)) }
+                "assign res" does jump("bodyExit") { writeRegister(getResultRegister(), registerUse(virB)) }
             }
 
         assertEquivalent(actualCFG, expectedCFG)
     }
+
+    // TODO: fix this test - accessing subfield makes whole structure land on stack
```

Comment:
You said that it could be an issue with variable analysis - usage of structure field `s.x` marks both `s` and `s.x` as used in our analysis (but not any other fields of `s`). I am not familiar with CFG generation well enough to know if that is a problem, but maybe you know?

Reply:
Its more issue of the Variable Allocation in the FunctionHandler I believe